core/src/main/java/io/grpc/Server.java

```diff
@@ -38,4 +41,57 @@
  * application code or interceptors.
  */
 @ThreadSafe
-public abstract class Server {}
+public abstract class Server {
```

Comment:
Yup, that makes sense.

@ejona86 @louiscryan @zhangkun83 @carl-mastrangelo @madongfly 

Folks, this PR expands the existing `Server` abstract class to contain the lifecycle methods from `ServerImpl`, so that we can use it with a generic builder, like we discussed for `Channel`.  Question is: what to call the interface (actually an abstract class) now: `Server`? `ManagedServer`? something else?  See discussion above.